BedrockServer.cpp

```diff
@@ -57,217 +68,346 @@ void BedrockServer_WorkerThread_ProcessDirectMessages(BedrockNode& node, Bedrock
-    // Initialize this thread
-    SInitialize();
-    SLogSetThreadPrefix("xxxxx ");
-    BedrockServer::Thread* data = (BedrockServer::Thread*)_data;
-    const SData& args = data->args;
-    bool readOnly = args.test("-readOnly");
-    BedrockServer::MessageQueue& queuedRequests = data->queuedRequests;
-    BedrockServer::MessageQueue& queuedEscalatedRequests = data->queuedEscalatedRequests;
-    BedrockServer::MessageQueue& processedResponses = data->processedResponses;
-    BedrockServer::MessageQueue& directMessages = data->directMessages;
-    SINFO("Starting " << (readOnly ? "read-only" : "read/write") << " worker thread for '" << data->name << "'");
+void BedrockServer::syncWorker(BedrockServer::ThreadData& data)
+{
+    // This needs to be set because the constructor for BedrockNode depends on it.
+    data.args["-worker"] = "false";
+    SInitialize(data.name);
+    SINFO("Starting sync thread for '" << data.name << "'");
 
     // Create the actual node
-    SINFO("Starting BedrockNode: " << args.serialize());
-    BedrockNode node(args, data->server);
+    SINFO("Starting BedrockNode: " << data.args.serialize());
+
+    // We let the sync thread create our journal tables here so that they exist when we start our workers.
+    int threads = max(1, data.args.calc("-readThreads"));
+    BedrockNode node(data.args, -1, threads, data.server);
+    _syncNode = &node;
     SINFO("Node created, ready for action.");
-    data->ready.set(true);
-
-    if (readOnly) {
-        for (;;) {
-            // Set the read-only node's state/master status coming from the replication thread.
-            // Only read-only nodes will allow an external party to set these properties.
-            SQLCState replicationState = data->replicationState.get();
-            const string masterVersion = data->masterVersion.get();
-            node.setState(replicationState);
-            node.setMasterVersion(masterVersion);
-
-            // Block until work is available.
-            fd_map fdm;
-            int maxS = queuedRequests.preSelect(fdm);
-            maxS = max(directMessages.preSelect(fdm), maxS);
-            S_poll(fdm, STIME_US_PER_S);
-            queuedRequests.postSelect(fdm);
-            directMessages.postSelect(fdm);
-
-            // If we've been instructed to shutdown and there are no more requests waiting
-            // to be processed, then exit the loop. Main thread will join us and continue
-            // the shutdown process.
-            bool shutdown = data->gracefulShutdown.get();
-            if (shutdown && queuedRequests.empty())
-                break;
 
-            // Process any direct messages from the main thread to us
-            BedrockServer_WorkerThread_ProcessDirectMessages(node, directMessages);
+    // Notify the parent thread that we're ready to go.
+    {
+        lock_guard<mutex> lock(_threadInitMutex);
+        _threadReady = true;
+    }
+    _threadInitVar.notify_all();
```

Comment:
Because std::condition_variable can't notify waiting threads on an atomic<bool>. From the docs:

> std::condition_variable works only with std::unique_lock<std::mutex>;